### DISCUSSION TITLE
How to rollback "Discard all changes"?

### DISCUSSION THREAD

**Discussion by antosha417** [jesseduffield/lazygit#2987] [Q&A]
More times that I care to admit I accidentally deleted newly created files with lazygit.
What do I do: 
1. create a bunch of new files while working on some awesome new feature.
2. now I'm ready to commit them. Usually I'm going over them in lazygit to scroll through diffs using `<c-d>` and then I hit `<space>` to stage those file.
3. and this is dangerous. Sometimes I press just `d` in place of `<c-d>` and then hit `space`. This discards changes to the entire file and I don't have any idea how to revert it. Since git does not yet track this file, it seems to be gone forever. 

Do you have any ideas how I can rollback this accidental deletions? Or even better avoid them in the first place

**Comment by jesseduffield** [comment]
Unfortunately there is no way to get that file back via git because git only has undo functionality relating to commits. It would be awesome if lazygit had a general undo feature that included changes to files but it would be a very complex feature.

As for preventing accidental discards, when you press d, a pop-up should appear confirming that you want to discard the file's changes. I consider that a pretty good protection.

**Reply by antosha417** [reply]
Yeah, you are completely right, pop-up should be enough. I'm just so used to it, that I press `<space>` automatically without even thinking. 
Can you please tell me how the file is deleted in this case?  Maybe if this file is not tracked by git we can move it to trash instead of deleting?

Also about pop-up I notices that when renaming the branch `<space>` and `y` are not considered as confirmation, only `<enter>` works. Maybe we can do the same for deleting the file? Or add `<space>` and `y` to branch renaming, just to have the same feel across application

**Reply by jesseduffield** [reply]
The file is deleted via golang's standard library which is equivalent to doing `rm`. So the file is not moved to the trash. At this point I don't plan on supporting moving the file to the trash because:
* looking on github, I can't see any popular go packages for that
* one package I found mentions performance issues: https://github.com/Bios-Marcel/wastebasket
* it doesn't solve the broader problem of discarding things accidentally e.g. discarding changes to an existing file. Solving the broader problem is more desirable and requires more thought. I'd be interested to see if there's some other open source git client out there that does this so we can see how they go about it.

As for space and y, the reason for that discrepancy is that when renaming a branch you're dealing with a prompt where you may actually want to include a space or 'y' as a character in the input. The reason we support space and y otherwise in non-prompt popups is because people have expressed a desire for it.
